object-api-multipart.go

```diff
@@ -0,0 +1,495 @@
+	if e != nil {
+		if e == errFileNotFound {
+			return false, nil
+		}
+		return false, e
+	}
+	return st.Mode.IsRegular(), nil
+}
+
+func (o objectAPI) PutObjectPart(bucket, object, uploadID string, partID int, size int64, data io.Reader, md5Hex string) (string, *probe.Error) {
+	// Verify if bucket is valid.
+	if !IsValidBucketName(bucket) {
+		return "", probe.NewError(BucketNameInvalid{Bucket: bucket})
+	}
+	if !IsValidObjectName(object) {
+		return "", probe.NewError(ObjectNameInvalid{Bucket: bucket, Object: object})
+	}
+	if status, e := o.isUploadIDExist(bucket, object, uploadID); e != nil {
+		return "", probe.NewError(e)
+	} else if !status {
+		return "", probe.NewError(InvalidUploadID{UploadID: uploadID})
+	}
+
+	partSuffix := fmt.Sprintf("%s.%d.%s", uploadID, partID, md5Hex)
+	fileWriter, e := o.storage.CreateFile(minioMetaVolume, path.Join(bucket, object, partSuffix))
+	if e != nil {
+		if e == errVolumeNotFound {
+			return "", probe.NewError(BucketNotFound{
+				Bucket: bucket,
+			})
+		} else if e == errIsNotRegular {
+			return "", probe.NewError(ObjectExistsAsPrefix{
+				Bucket: bucket,
+				Prefix: object,
+			})
+		}
+		return "", probe.NewError(e)
+	}
+
+	// Initialize md5 writer.
+	md5Writer := md5.New()
+
+	// Instantiate a new multi writer.
+	multiWriter := io.MultiWriter(md5Writer, fileWriter)
+
+	// Instantiate checksum hashers and create a multiwriter.
+	if size > 0 {
+		if _, e = io.CopyN(multiWriter, data, size); e != nil {
+			fileWriter.(*safe.File).CloseAndRemove()
+			return "", probe.NewError(e)
+		}
+	} else {
+		if _, e = io.Copy(multiWriter, data); e != nil {
+			fileWriter.(*safe.File).CloseAndRemove()
+			return "", probe.NewError(e)
+		}
+	}
+
+	newMD5Hex := hex.EncodeToString(md5Writer.Sum(nil))
+	if md5Hex != "" {
+		if newMD5Hex != md5Hex {
+			fileWriter.(*safe.File).CloseAndRemove()
+			return "", probe.NewError(BadDigest{md5Hex, newMD5Hex})
+		}
+	}
+	e = fileWriter.Close()
+	if e != nil {
+		return "", probe.NewError(e)
+	}
+	return newMD5Hex, nil
+}
+
+func (o objectAPI) ListObjectParts(bucket, object, uploadID string, partNumberMarker, maxParts int) (ListPartsInfo, *probe.Error) {
+	// Verify if bucket is valid.
+	if !IsValidBucketName(bucket) {
+		return ListPartsInfo{}, probe.NewError(BucketNameInvalid{Bucket: bucket})
+	}
+	if !IsValidObjectName(object) {
+		return ListPartsInfo{}, probe.NewError(ObjectNameInvalid{Bucket: bucket, Object: object})
+	}
+	if status, e := o.isUploadIDExist(bucket, object, uploadID); e != nil {
+		return ListPartsInfo{}, probe.NewError(e)
+	} else if !status {
+		return ListPartsInfo{}, probe.NewError(InvalidUploadID{UploadID: uploadID})
+	}
+	result := ListPartsInfo{}
+	marker := ""
+	nextPartNumberMarker := 0
+	if partNumberMarker > 0 {
+		fileInfos, _, e := o.storage.ListFiles(minioMetaVolume, path.Join(bucket, object, uploadID)+"."+strconv.Itoa(partNumberMarker)+".", "", false, 1)
```

Comment:
This is a problem of current multipart approach. Since we dont keep any metadata related to previous parts. So deleting a exact previous part would be really tricky as we need to know the md5sum of previous part.